### PR TITLE
test/new-e2e/tests/apm: add first apm tests

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

AIT-9391

Implement `TraceAgentMetrics` in `new-e2e`. The tests ensure the trace-agent submits its `datadog.trace_agent.*` metrics correctly through dogstatsd. The tests relies on the metrics fake intake.

Implement `TracesOnUDS` and `StatsOnUDS`, the tests ensure the agent produces the expected traces and stats payloads for traffic coming on UDS.

### Motivation

Tests are good.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: .gitlab/e2e.yml
Comment (ahmed-mez) [inline .gitlab/e2e.yml:325]: These are copy/pasted from the existing jobs - can the reviewers please confirm it's correct?

**Review by knusbaum — APPROVED** [review]
LGTM. I put tests in `test/new-e2e/tests/trace-agent`, which looks like it's how the other daemons named theirs, but I don't think it really matters. I'll leave it up to you.

File: .gitlab/e2e.yml
Comment (pducolin) [reply .gitlab/e2e.yml:335]: Required if you have tests based on the agent docker image (docker, kindvm, ecs or eks). If you're using the agent on host, this is not needed.

File: test/new-e2e/tests/apm/docker_test.go
Comment (pducolin) [inline test/new-e2e/tests/apm/docker_test.go:31]: **💭 thought**
Note: docker support is currently broken as VMs cannot access the images that we publish to the `agent-qa` ECR. @KevinFairise2 is working on it, should be done by end of next week.

File: test/new-e2e/tests/apm/docker_test.go
Comment (pducolin) [inline test/new-e2e/tests/apm/docker_test.go:51]: **💭 thought**
Is it possible that the trace-agent starts before the fakeintake is ready ? We should wait for the fakeintake to be ready, but if this is flaky, it sounds like a race condition between trace-agent start and fakeintake readiness

File: test/new-e2e/tests/apm/docker_test.go
Comment (pducolin) [inline test/new-e2e/tests/apm/docker_test.go:156]: **💭 thought**
This should be doable on the docker client, that uses docker sdk. We use it to run command on containers deployed within a docker compose https://github.com/DataDog/datadog-agent/blob/main/test/new-e2e/pkg/utils/e2e/client/docker.go#L56-L60

**Review by pducolin — APPROVED** [review]
Nice !

File: .gitlab/e2e.yml
Comment (ahmed-mez) [reply .gitlab/e2e.yml:335]: Thanks, I'm actually using the agent on docker in these tests.

File: test/new-e2e/tests/apm/docker_test.go
Comment (ahmed-mez) [reply test/new-e2e/tests/apm/docker_test.go:31]: Do you think it's a blocker for merging the PR?

File: test/new-e2e/tests/apm/docker_test.go
Comment (ahmed-mez) [reply test/new-e2e/tests/apm/docker_test.go:51]: That's a possibility, I'm also suspecting a more subtle bug like a race between the core-agent and the trace-agent, since these are dogstatsd metrics sent by the trace-agent through the core-agent. I'll file a bug to investigate this further.

Comment (pducolin) [reply test/new-e2e/tests/apm/docker_test.go:51]: To exclude the first, you can dual-ship to `dddev` and see if you get it there (for the investigation)

File: test/new-e2e/tests/apm/docker_test.go
Comment (ahmed-mez) [reply test/new-e2e/tests/apm/docker_test.go:156]: Thanks! It looks like `ExecuteCommand` does the equivalent of `docker exec` in a given container, and cannot be used to run new containers which is what we're doing here.
My plan (as written in the TODO) is to define the fake workload in a docker compose and provide it with `WithExtraComposeManifest`. Lmk if it's a bad idea 😅

File: test/new-e2e/tests/apm/docker_test.go
Comment (pducolin) [reply test/new-e2e/tests/apm/docker_test.go:31]: Unfortunately yes :(. You could add it and allow to fail it or add it and skip it with the message ("blocked by APL-2786")

Comment (pducolin) [reply test/new-e2e/tests/apm/docker_test.go:31]: It runs fine locally, it will fail on main as we use the agent image published to the `agent-qa` ECR

Comment (ahmed-mez) [reply test/new-e2e/tests/apm/docker_test.go:31]: Will do, thanks!

Comment (pducolin) [reply test/new-e2e/tests/apm/docker_test.go:31]: Will add the suggestion where this should happen

File: test/new-e2e/tests/apm/docker_test.go
Comment (ahmed-mez) [reply test/new-e2e/tests/apm/docker_test.go:156]: We'll have to do it in the next task once we start tweaking the fake workload config.

File: test/new-e2e/tests/apm/docker_test.go
Comment (pducolin) [inline test/new-e2e/tests/apm/docker_test.go:29]: **💬 suggestion**
```suggestion
	isCI, _ := strconv.ParseBool(os.Getenv("CI"))
	if isCi {
		t.Skipf("blocked by APL-2786")
	}
	devModeEnv, _ := os.LookupEnv("E2E_DEVMODE")
```
Requires adding `strconv` and `os` to imports too

**Review by knusbaum — APPROVED** [review]

File: test/new-e2e/tests/apm/docker_test.go
Comment (knusbaum) [reply test/new-e2e/tests/apm/docker_test.go:160]: I actually like the inline method here, since it's super clear and in the same code as the test being run.

**Review by sblumenthal — APPROVED** [review]

**Pull request merged**
